### ISSUE TITLE
Composite escape handler doesn't check if another non escape character was typed before 

### ISSUE THREAD

**Issue by asvetliakov** [asvetliakov/vscode-neovim#330]
@theol0403 that's another issue, second composite key handler doesn't check if you've typed another character between first & second escape keys. Should be easy to fix

_Originally posted by @asvetliakov in https://github.com/asvetliakov/vscode-neovim/issues/324#issuecomment-680853589_

**Comment by chaorace** [comment]
Reposting my concern here since it has a dedicated issue now:

> I question if the current behavior is even consistent with vim's chording mechanism. Wouldn't it make more sense to reject a 'jk' key chord if it's interrupted instead of trying to salvage the interrupting keystrokes?

**Comment by ChasKane** [comment]
Insert mode is handled by vscode. The only time vscode communicates with the neovim instance during insert mode is when a command is executed due to a key sequence registered in `keybindings.json`. Thus, the [composite escape key command code I modified](https://github.com/vscode-neovim/vscode-neovim/pull/892) is only called for the key bindings made by the user, not all key strokes in insert mode.

If the neovim extension isn't notified of all key strokes, I don't think there's a way for neovim to decide whether to issue an `esc` key based on whether keys have been pressed between `j` and `k`. the new compositeEscape code stores the last key that was pressed in insert mode which is registered in `keybindings.json` for the same command, not any other keystrokes.

As far as I know, there's no way to notify neovim of all keystrokes unless you register a command for every key, which would be inelegant and defeat the purpose of having vscode handle insert mode. I don't know if that would lead to performance problems either, or, frankly, if even registering a command for just 2 keys, as in the case of `jk`, would slow it down under any unusual circumstances.

**Comment by theol0403** [comment]
> Insert mode is handled by vscode. The only time vscode communicates with the neovim instance during insert mode is when a command is executed due to a key sequence registered in `keybindings.json`. Thus, the [composite escape key command code I modified](https://github.com/vscode-neovim/vscode-neovim/pull/892) is only called for the key bindings made by the user, not all key strokes in insert mode.
> 
> If the neovim extension isn't notified of all key strokes, I don't think there's a way for neovim to decide whether to issue an `esc` key based on whether keys have been pressed between `j` and `k`. the new compositeEscape code stores the last key that was pressed in insert mode which is registered in `keybindings.json` for the same command, not any other keystrokes.
> 
> As far as I know, there's no way to notify neovim of all keystrokes unless you register a command for every key, which would be inelegant and defeat the purpose of having vscode handle insert mode. I don't know if that would lead to performance problems either, or, frankly, if even registering a command for just 2 keys, as in the case of `jk`, would slow it down under any unusual circumstances.

Youre right. I'm considering making a PR that registers the type command at all times, so then you would have access to the typed keys, and it might not cause issues with the new experimentalAffinity setting. This would be a step towards full insert-mode integration, but not all the way there. 

Another possibility is to use a vscode api to check what the letter before the cursor is in the jk handler.

**Comment by BlueDrink9** [comment]
I think #886 being merged is sufficient to close this issue?
